File: Problems/No.1461_Check_If_a_String_Contains_All_Binary_Codes_of_Size_K/check_if_a_string_contains_all_binary_codes_of_size_k.py
class Solution(object):
    def hasAllCodes(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: bool
        """

        # parameters
        k_binary_codes = set()
        all_codes = 1<<k

        # start computing
        if (len(s) < k+all_codes-1):
            return False
        for i in range(len(s)-k+1):
            k_binary_codes.add(s[i:i+k])
        
        return True if (len(k_binary_codes) == all_codes) else False

File: Problems/No.1461_Check_If_a_String_Contains_All_Binary_Codes_of_Size_K/check_if_a_string_contains_all_binary_codes_of_size_k.cpp
class Solution {
public:
    bool hasAllCodes(string s, int k) {
        
        // parameter
        std::set<std::string> k_binary_codes;
        int all_codes = 1<<k;

        // start computing
        if (s.length() < k+all_codes-1) return false;
        for (int i=0; i<=s.length()-k; i++)
            k_binary_codes.insert(s.substr(i, k));
        
        return (k_binary_codes.size() == all_codes) ? true : false;

    }
};
